### PR TITLE
Bug 1235399 - enable tc-proxy to work with certificates

### PR DESCRIPTION
**Pull request by petemoore**
Not yet tested...

### CODE REVIEW

**Comment by petemoore** [comment]
In order to test this in travis, we'll need to implement http://docs.taskcluster.net/auth/temporary-credentials/ in go - that hasn't been done yet, but should live in [taskcluster-client-go](http://taskcluster.github.io/taskcluster-client-go).

**Comment by petemoore** [comment]
Note, I logged in just now with login.taskcluster.net, and got myself some temporary credentials. I then used these for running the tests, and all the tests passed. \o/

We should still do this as part of the tests - but just so you know, it does seem to be working, even if we don't have unit tests yet.

**Comment by petemoore** [comment]
Note - the travis failure is only because coverage has reduced overall with this change. When I add tests for temp creds, this will obviously be resolved.

File: README.md [hunk 1]
```diff
@@ -11,7 +11,9 @@ scheduler) without hardcoding credentials into the containers
 themselves.
 
 Credentials are expected to be passed via the `TASKCLUSTER_CLIENT_ID`
-and `TASKCLUSTER_ACCESS_TOKEN` environment variables.
+and `TASKCLUSTER_ACCESS_TOKEN` environment variables, optionally with
+a `TASKCLUSTER_CERITIFICATE` environment variable in the case of using
+temporary credentials.
```
Comment (djmitche) [inline README.md]: Note that at least with docker, you do NOT want to communicate these with env vars, as those vars get injected into every linked container (including the task container).  Docker-worker uses the command-line options instead.

File: main.go [hunk 2]
```diff
@@ -59,7 +60,12 @@ func main() {
 		accessToken = os.Getenv("TASKCLUSTER_ACCESS_TOKEN")
 	}
 
-	log.Printf("%v - %v", clientId, accessToken)
+	certificate := arguments["--certificate"]
+	if certificate == nil || certificate == "" {
```
Comment (djmitche) [inline main.go]: `certificate.(string) == ""`?

File: taskcluster/authorization.go [hunk 3]
```diff
@@ -17,17 +17,21 @@ type delegationOptions struct {
...
 	credentials := &hawk.Credentials{
 		ID:   clientId,
 		Key:  accessToken,
 		Hash: sha256.New,
 	}
-	return hawk.NewRequestAuth(http, credentials, 0)
+	auth := hawk.NewRequestAuth(http, credentials, 0)
+	if certificate != "" {
+		auth.Ext = base64.StdEncoding.EncodeToString([]byte("{\"certificate\":" + certificate + "}"))
+	}
```
Comment (djmitche) [inline taskcluster/authorization.go]: Maybe use a helper for this, since it's formatting-sensitive and repeated below?

**Comment by djmitche** [comment]
Looks good :)

File: main.go (hunk 2 above)
Comment (petemoore) [inline main.go]: Not needed, see: https://play.golang.org/p/XNLCnLGNoT

**Comment by petemoore** [comment]
Although the latest commit appears like a big refactor, I essentially just inlined the tests as a local function, and then call it with both permanent credentials and temporary credentials. Due to the increased indentation, the diff looks like quite a mess, but essentially that's all I did...

**Comment by petemoore** [comment]
@djmitche can you take a look again? Thanks!

File: taskcluster/authorization_test.go [hunk 4]
```diff
@@ -28,143 +32,167 @@ var (
...
-	defer resp.Body.Close()
-	json, err := readJson(resp)
+func testWithPermCreds(t *testing.T, test IntegrationTest) {
+	skipIfNoPermCreds(t)
+	test(t, permCredentials)
+}
 
+func testWithTempCreds(t *testing.T, test IntegrationTest) {
+	skipIfNoPermCreds(t)
+	tempCredentials, err := permCredentials.CreateTemporaryCredentials(1*time.Hour, "*")
```
Comment (petemoore) [inline taskcluster/authorization_test.go]: Whoops - I forgot to reduce this from '*' - I still need to do that...

**Comment by djmitche** [comment]
Looks good

**Comment by petemoore** [comment]
I updated the test - and discovered a bug in the process! In short, currently AuthorizationDelegate function is not working with temporary credentials - the easiest way to fix this will be to use tcclient library instead.

**Comment by petemoore** [comment]
Fixed the genuine bug that was discovered via a test. With the updates, taskcluster-proxy is now using the same tcclient library as taskcluster-client-go uses under the hood, so we now have less code to manage

@djmitche hopefully this should be the last review! Thanks.

**Comment by djmitche** [comment]
:thumbsup:

**Pull request merged**